### PR TITLE
WIP: Return redirect url on xhr requests

### PR DESCRIPTION
**Pull request by dleve123**
## Problem

We don't want to require users to sign in twice when accessing an auction (once for www.artsy.net, another time when accessing live.artsy.net). As such, we want to be able to sign into Gravity when signing into Force. At some point, this behavior worked (https://github.com/artsy/artsy-passport/pull/80), but was broken. We strongly suspect that this broke when we migrated Force's auth to use modals instead of server-side ajax.

To effectively authenticate with Gravity from Force, we need to expose the ability to authenticate with Force **client-side**. This PR does that by leveraging existing Gravity endpoints that perform authentication of `GET`s.

A sibling Force PR leverages the JSON payload now returned from artsy-passport to perform the client side redirect: https://github.com/artsy/force/pull/4349

## Solution 

Include a `api_sign_in_with_redirect` in the JSON response returned when auth'ing to Force over AJAX.

Jira: https://artsyproduct.atlassian.net/browse/AUCT-281

---

TODO:

- [x] Figure out if we need the [short circuit return](https://github.com/artsy/artsy-passport/compare/master...dleve123:return-redirect-url-on-xhr-requests?expand=1#diff-c7d8f6e9781b93e5f26d9a3fbfbf13c0R125) it makes it hard to test locally
  - opted to remove for now: https://github.com/artsy/artsy-passport/pull/133#discussion_r308912136
- [x] Resolve test failures
- [x] Bump minor version

### CODE REVIEW

File: lib/app/lifecycle.coffee [hunk 1]
```diff
@@ -119,16 +119,21 @@ crypto = require 'crypto'
...
 @ssoAndRedirectBack = (req, res, next) ->
-  return res.send { success: true, user: req.user.toJSON() } if req.xhr
-  parsed = parse redirectBack req
-  parsed = parse resolve opts.APP_URL, parsed.path unless parsed.hostname
-  domain = parsed.hostname?.split('.').slice(1).join('.')
-  return redirectBack(req, res) if domain isnt 'artsy.net'
+  urlToRedirectTo = parse redirectBack req
+  urlToRedirectTo = parse resolve opts.APP_URL, urlToRedirectTo.path unless urlToRedirectTo.hostname
+  domain = urlToRedirectTo.hostname?.split('.').slice(1).join('.')
+  # return redirectBack(req, res) if domain isnt 'artsy.net'
```
Comment (dleve123) [inline lib/app/lifecycle.coffee]: This line was introduced in https://github.com/artsy/artsy-passport/pull/80/commits/f0c51b8addc89d4ab2111c49517a77d119149344 (PR: https://github.com/artsy/artsy-passport/pull/80) by Craig.

Does anyone recall the rational for this short-circuit return? It makes it harder to develop against artsy-passport, so I would prefer to remove it if possible.

cc: @mzikherman @sweir27 (who created some PRs which referenced #80)

**Comment by dleve123** [comment]
No CI for this project, so pasting evidence of a passing test suite:

```
dl@phoenix:~/src/artsy-passport (return-redirect-url-on-xhr-requests)
$ git show HEAD
commit 7e0ea4fc270fc3ee221d9710d82c48196aaa862a (HEAD -> return-redirect-url-on-xhr-requests)
Author: Daniel Levenson <dleve123@gmail.com>
Date:   Tue Jul 30 15:35:00 2019 -0400

    Set the request_uri to the request’s referrer.
    
    This commit also organizes test/app/lifecycle.coffee a bit more with
    `context`s and `describes`.
    
    Before this change, the redirect_uri only matched on `redirect-to`-like
    indicators in the request query string or payload.
    
    This should *not* be a breaking change to any non-AJAX clients because
    we return the `request.referral` _only if_ there are no other
    `redirect-to`-like indicators.
    
    However, to handle the case where we want auth with Gravity from an AJAX
    request, the redirect_uri that we would want to expose to Gravity is the
    referrer of the request. Example data flow:
    
    1. GET www.artsy.net/auctions
    
    (User opens modal and submits valid credentials, which leads to)
    
dl@phoenix:~/src/artsy-passport (return-redirect-url-on-xhr-requests)
$ yarn test
yarn run v1.17.3
$ mkdir -p dist/app && cp lib/app/sanitize_redirect.js dist/app && mocha test/app && mocha test/passport


  analytics
    ✓ tracks signup (102ms)
    ✓ tracks login
    ✓ passes along modal_id and acquisition_initiative submitted fields
    ✓ passes along acquisition_initiative query params for OAuth links
    ✓ doesnt hold on to the temporary session variable

  lifecycle
    #onLocalLogin
      when successful
        ✓ authenticates locally and passes on (173ms)
        ✓ authenticates locally and passes on
      when erroring
        ✓ redirects invalid passwords to login
        ✓ sends generic error to the error handler
    #onError
      ✓ Nexts on error
    #onLocalSignup
      - creatse a user
      - suggests an email if its invalid and redirects back to the signup page
      - sends 500s as json for xhr requests
      ✓ passes the recaptcha_token through signup
      ✓ passes the user agent through signup (71ms)
    #beforeSocialAuth
      ✓ sets session redirect
      ✓ sets the session to skip onboarding
    #afterSocialAuth
      ✓ doesnt redirect to personalize if skip-onboarding is set
      ✓ surfaces blocked by facebook errors
      ✓ passes random errors to be rendered on the login screen
      with an error
        - redirects back to the login page and explains the account was previously linked
        - redirects to the settings page and explains if the account is linked to another account
        - passes unknown errors
      when successful
        - redirects to settings if linking
        - redirects to the personalize page if signing up
        - reidrects back if logging in
    #ensureLoggedInOnAfterSignupPage
      ✓ redirects to the login page, and back, without a user
    #ssoAndRedirectBack
      when the incoming request is server-side (not XHR)
        ✓ redirects signups to personalize
        ✓ doesnt redirect to personalize if skipping onboarding
        ✓ single signs on to gravity
      when the incoming request is an XHR
        ✓ returns JSON with success: true
        the api_sign_in_with_redirect json value url
          ✓ includes the trust token returned from the api
          ✓ includes a redirect_uri equal to the request referrer

  locals
    ✓ escapes the error message html (56ms)
    - sets the user if logged in
    - sets csrf to sharify
    - sets the query param error for login & signup pages
    - provides the artsy passport options

  #logout
    - logs out, deletes the auth token, and redirects home
    - sets the access token

  #denyBadLogoutLinks
    ✓ ensures the referrer is artsy to avoid the img src= logout hack

  sanitizeRedirect
    ✓ lets artsy.net through
    ✓ lets artsy.net without a protocol through
    ✓ lets artsy.net subdomains through
    ✓ lets artsy.net subdomains without a protocol through
    ✓ lets relative paths through
    ✓ lets deeply nested artsy.net subdomains through
    ✓ lets artsy.net links through
    ✓ lets localhost through
    ✓ lets internal paths through
    ✓ blocks data urls
    ✓ blocks external links not whitelisted; redirects to root
    ✓ blocks external links that lack protocol; redirects to root
    ✓ blocks other protocols; redirects to root
    ✓ blocks other protocols; redirects to root
    ✓ blocks malformed URLs (1)
    ✓ blocks malformed URLs (2)
    ✓ blocks malformed URLs (3)
    ✓ strips out whitespace

  token login middleware
    #headerLogin
      ✓ logs in a user if they pass their access token as a header (174ms)
      ✓ does not log in a user on sign out
    trustTokenLogin
      ✓ immediately nexts if there is no trust_token query param
      ✓ logs the user in when there is a trust_token present, redirecting to a url sans trust_token param
      ✓ preserves any other query string params
      ✓ nexts on failed code response


  50 passing (778ms)
  15 pending



  passport callbacks
    ✓ get a user with an access token email/password (101ms)
    ✓ get a user with an access token facebook
    ✓ passes the user agent through login
    ✓ passes the user agent through facebook signup

  #serialize
    ✓ only stores select data in the session (86ms)
    ✓ add authentications
    ✓ works when theres an error from Gravity

  #deserialize
    ✓ wraps the user data in a model


  8 passing (250ms)

```

**Comment by dleve123** [comment]
This PR still needs a version bump, but would love review from reviewers here! Assigned to myself b/c not ready to merge yet (but I am definitely going to wait for many reviewers here before integrating)

File: lib/app/lifecycle.coffee (hunk 1 above)
Comment (joeyAghion) [reply lib/app/lifecycle.coffee]: Seems like it might be security-related but I don't fully understand. If you do drop this line, the line above can go away as well.

File: lib/app/lifecycle.coffee [hunk 2]
```diff
@@ -119,16 +119,21 @@ crypto = require 'crypto'
...
-      res.redirect "#{opts.ARTSY_URL}/users/sign_in" +
+      return res.redirect urlToRedirectTo.href if err
+
+      apiAuthWithRedirectUrl = "#{opts.ARTSY_URL}/users/sign_in" +
         "?trust_token=#{sres.body.trust_token}" +
-        "&redirect_uri=#{parsed.href}"
+        "&redirect_uri=#{urlToRedirectTo.href}"
+
+      if req.xhr?
+        res.send { success: true, api_sign_in_with_redirect: apiAuthWithRedirectUrl, user: req.user.toJSON() }
```
Comment (joeyAghion) [inline lib/app/lifecycle.coffee]: If I understand correctly, this gives clients the opportunity to navigate their user through the api and back to the current application. Isn't the client already able to do that though, since it specifies the redirect target in the first place?

Comment (dleve123) [reply lib/app/lifecycle.coffee]: Can you elaborate? Not really following you here.

**Comment by dleve123** [comment]
Success CI build: https://circleci.com/gh/artsy/artsy-passport/83

File: lib/app/lifecycle.coffee (hunk 1 above)
Comment (sweir27) [reply lib/app/lifecycle.coffee]: I can't recall anything specific about this (@dleve123 do you have those PRs you mention handy?). It does appear to be security-related, but I suppose we could add some kind of environment-specific condition to avoid the developer experience issue.

That being said, I'm not sure when we'd encounter this situation?

Comment (dleve123) [reply lib/app/lifecycle.coffee]: > (@dleve123 do you have those PRs you mention handy?)

https://github.com/artsy/prediction/pull/208, https://github.com/artsy/microgravity-private/pull/1401 (you didn't create them, just were assigned to them, my b)

> It does appear to be security-related, but I suppose we could add some kind of environment-specific condition to avoid the developer experience issue.

So both of this repo and force is open source  – perhaps there's a vector where someone could run a local force, and, given a valid set of auth credentials to gravity (which isn't open source), get auth'ing redirect link to gravity? Either way, it's really straightforward to spoof being on the artsy domain (there are in fact steps for this in this project's [README](https://github.com/artsy/artsy-passport/blob/master/README.md#contributing)).

Comment (mzikherman) [reply lib/app/lifecycle.coffee]: I believe the intent is to only sign someone in to gravity, if they are logging in to Force (or any other Artsy app, which uses artsy-passport).

Theoretically this lib is OSS so you might use this for your own client application (not an Artsy app, not on artsy.net). In that case, you probably wouldn't want this (since the only convenience is to be able to login to _other_ apps).

I don't know if it's security-related per se, and more just expected behavior.

** Caveat- this is just my speculation!

File: lib/app/lifecycle.coffee (hunk 2 above)
Comment (mzikherman) [reply lib/app/lifecycle.coffee]: I _think_ what @joeyAghion might be saying (correct me if I'm wrong!), is that, rather than encode this behavior in this lib (being able to provide the redirect location for a Gravity auth), maybe it can just stay in the client (Force)?

According to the [dependency graph](https://github.com/artsy/artsy-passport/network/dependents), it doesn't actually look like there are any other active clients besides Force at the moment. So if it's not realistic/feasible for _other_ clients to take advantage of this, it might be easier to just work in Force (or Reaction, if that's where authentication flow/modal stuff lives), and avoid faffing with this library. Those are more modern repos and technologies that are better supported (personally I view some of Artsy's older libs/repos, such as this one, as pseudo-deprecated, in that there's usually better options _today_, and these tend to rustle jimmies).

Comment (dleve123) [reply lib/app/lifecycle.coffee]: > According to the dependency graph, it doesn't actually look like there are any other active clients besides Force at the moment. 

This library is being used by [artsy/vanity](https://github.com/artsy/vanity/blob/master/lib/setup.coffee#L11) as well.

>  So if it's not realistic/feasible for other clients to take advantage of this, it might be easier to just work in Force (or Reaction, if that's where authentication flow/modal stuff lives), and avoid faffing with this library. Those are more modern repos and technologies that are better supported (personally I view some of Artsy's older libs/repos, such as this one, as pseudo-deprecated, in that there's usually better options today, and these tend to rustle jimmies).

So a few things:

1. I'll acknowledge by bias here as someone that spent a significant amount of time learning up on cofeescript and related technologies here. I'm interested in solving this problem, and moving onto other problems :)

2. Is there a concrete plan to move off of artsy-passport? If so, I'm all for investing in solving this problem via a more supported solution path.

However, if there isn't, then I don't want **avoid** solution-ing within artsy-passport in this way because there's more modern technology other places. Moreover, there already seems to be a strong separation of concerns between "authentication" logic in artsy-passport and other view/route/business logic within force/reaction, and I think there's value in that code boundary.

---

If there are strong feelings on another concrete ways to solution here, I'm all ears. I'd appreciate a pair on it though!

I would like to goal set to have concrete next steps here by end of day (EST) tomorrow though.

cc/ @mzikherman @joeyAghion

Comment (joeyAghion) [reply lib/app/lifecycle.coffee]: To clarify, yes, my question was simply: does artsy-passport _need_ to offer this redirect pattern (e.g., because otherwise some value such as an internal token wouldn't be accessible), or could a client decide to perform this redirect on its own?

This library is definitely in active use so I don't see any reason to _avoid_ updating it. I only asked in case we didn't _need_ to change this API for a particular client use case.

Comment (eessex) [reply lib/app/lifecycle.coffee]: In case these haven't been surfaced yet, there are some docs addressing redirects here: https://github.com/artsy/reaction/blob/c9d18e7f9ff47f2bf72f2f8f950d345d976d9d20/docs/authentication.md#redirects

Is the original form implementation using either `destination` or `redirectTo`, and we've confirmed it does not work? https://github.com/artsy/reaction/blob/20d28d5628c680270dfe01b4875a7732fd7a9576/src/Components/Authentication/Types.ts#L63

File: lib/app/lifecycle.coffee (hunk 1 above)
Comment (dleve123) [reply lib/app/lifecycle.coffee]: Got it, thanks for the context all! Don't want to dedicate a ton of mind-share to this and seems like we don't have a great reason to keep this limitation in, so opting to remove this LOC. FWIW, `yarn test` passes without it.

File: lib/app/lifecycle.coffee (hunk 2 above)
Comment (dleve123) [reply lib/app/lifecycle.coffee]: > In case these haven't been surfaced yet, there are some docs addressing redirects here: https://github.com/artsy/reaction/blob/c9d18e7f9ff47f2bf72f2f8f950d345d976d9d20/docs/authentication.md#redirects

I've checked it out before, but thanks for the reminder!

> Is the original form implementation using either destination or redirectTo, and we've confirmed it does not work? https://github.com/artsy/reaction/blob/20d28d5628c680270dfe01b4875a7732fd7a9576/src/Components/Authentication/Types.ts#L63

Yup, confirmed it doesn't work, if by work, we mean create a session with Gravity. Check out [this GH comment](https://github.com/artsy/force/pull/4349#issuecomment-516882362) for more information.

WRT `redirectTo` or `destination` – my understanding of the code is that when calling [`ModalManager` in Force](https://github.com/artsy/force/blob/33ad0bf9f13c737a9e6993cf1bf9febc4f70c0b4/src/desktop/apps/authentication/components/ModalContainer.tsx#L75), we _always_ pass a `handleSubmit` prop, the value of which is the function defined in [`helpers.ts` which first considers the `redirectTo` of the modal options or respond with default logic](https://github.com/artsy/force/blob/79e11ee82200a990240f5a1b0e6a3fcd45dbba0e/src/desktop/apps/authentication/helpers.ts#L78).

To the best of my knowledge, there's only 1 place in force when `redirectTo` is set to a value that isn't the current window location, when registering for an auction.

Screen Shot 2019-08-01 at 11 43 18 PM

So one limitation here when logging in and registering at the same time for an auction, we won't be creating a session with Gravity.

---

To see how we could mitigate against the limitation mentioned above, I'm going to give implementing this logic within Force a stab by providing some form of wrapper around `afterSubmit`.

**Comment by dleve123** [comment]
Marking this as a draft so that we don't merge as I'm trying out a solution within Force per https://github.com/artsy/artsy-passport/pull/133#discussion_r309974323

**Comment by dleve123** [comment]
Closing in favor of https://github.com/artsy/force/pull/4379
